packages/react-query/src/__tests__/useQueries.test.tsx

```diff
@@ -84,15 +84,15 @@ describe('useQueries', () => {
         queries: [
           {
             queryKey: [key1, count],
-            keepPreviousData: true,
+            placeholderData: (previousData?: number) => previousData,
```

Comment:
is the type for `previousData` not inferred here?

Reply:
This is potentially some kind of bug in TS.
When you check the type of `placeholderData`, it shows the proper type with `number` inferred.
But marks the function parameter as `any`, like it would lost the inferred context somehow.
Tried updating TS to latest with no success.

![image](https://user-images.githubusercontent.com/28151934/209805189-d5657cf9-c047-45c5-bce5-8c3030f095ce.png)

Reply:
seems like this is specific to `useQueries` only? Maybe @artysidorenko can give us some help here :)

Reply:
Hi - yes unfortunately it's a limitation that's been appearing on all the `useQueries` callback parameters from the beginning ([sorta-vaguely-documented here in the tests](https://github.com/TanStack/query/blob/main/packages/react-query/src/__tests__/useQueries.test.tsx#L674-L717)). Similar to what you saw: it's really strange how the IDE/intellisense suggests the correct type on hover, but it still complains about `no-implicit-any`.

I wasn't really sure why it's happening, but you've motivated me to put together a simplified example and share it as a [TS issue](https://github.com/microsoft/TypeScript/issues/52047) - hopefully will provide some insight 🤞

Reply:
thanks @artysidorenko - I've subscribed to this issue. Nothing more we can do here I think for now.

Reply:
The reported issue can be implemented using so-called "reverse mapped types". You can see the solution to the reported issue [here](https://github.com/microsoft/TypeScript/issues/52047#issuecomment-1367642953)

It would be fun for me to play around with your types in an attempt to make them work. I'm afraid though that your types are quite complex and I don't have time to figure out what they currently do on my own. Would somebody from your team like to jump on a call with me and spike this someday?

Reply:
I would love to, but I don't think I understand those types enough to be of much help. @artysidorenko is the wizard who wrote them. I'd love to be a fly on the wall when you're talking about it 😁

Reply:
I concluded that the main problem here, for your real thing, is that you have multiple "generics" per a tuple element. Unfortunately, reverse mapped types can't handle this too well.

They can only infer directly into `T[K]` - but when we try to treat the whole tuple element as what we try to infer then we end up with the original problem as we can't relate different properties of what we have inferred as `T[K]` as easily.

I tried quite a bit to provide the argument's type for the `handler` in a different way. My main bets were different kinds of intersections*, other mapped types wrapping the argument position, and even circular constraints for `T`. None of those turned out to work here.

I somewhat strongly believe that this improvement would make it quite trivial to implement: https://github.com/microsoft/TypeScript/issues/51612

*I even tried to intersect multiple reverse mapped types. This didn't work out but it feels like **maybe** it wouldn't be that hard to make it work in the compiler. Such things often don't work because more advanced logic is missing intersections/unions handling and adding that is often not that hard.

Reply:
Hi @Andarist, thanks for looking into this 🙏 

> the main problem here, for your real thing, is that you have multiple "generics" per a tuple element

Yep, indeed, this is what I kept running into. In the end abandoned mapped types entirely, and replaced with [that bulky utility type `QueriesOptions`](https://github.com/TanStack/query/blob/9d9aea5fb12eb89dec54c619845b3d226b53cf2b/packages/react-query/src/useQueries.ts#L98), to walk through the argument array recursively and infer each of those "multi-generic" tuple elements, one at a time. It sort-of does the job, but as you notice it's not great readability-wise, there's a fair few sticking plasters to cover edge cases, and some downright weirdness, like the no-implicit-any issue here.

Let me know if you want to discuss further - my coverage isn't great at the moment (on vacation - coming to you from a hotel lobby in Mexico) but back home and more responsive from 13th Jan.

Reply:
I think I'm going to continue poking around with those reverse mapped types in the compiler. I have 2 ideas on how this could be solved and I will attempt to implement both of them in the coming months. Stay tuned 😉

Reply:
I **think** that i’ve found a patch for the compiler that would allow is to rewrite those types and make them easier and make those function arguments inferrable.

Let’s schedule this call to spike a potential rewrite 🚀

Reply:
@Andarist @artysidorenko let me know if I can help here. Which time would work for you ? I'm thinking that Discord might be a better way to communicate. Feel free to contact me there

Reply:
Sounds good - pinged you on discord just now.

Also keeping an eye on andarist's [TS PR currently in code review](https://github.com/microsoft/TypeScript/pull/52062). Looks like [the example given in the PR test](https://github.com/microsoft/TypeScript/pull/52062/files#diff-d9f6d4a38129503d13ff5d4c592a3402d123eea76e86254a5479ffebdc50b909) fits the useQueries use case pretty closely 🤞 once that's merged I think we'll be unblocked here.